staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queuewait/queuewait.go

```diff
@@ -0,0 +1,69 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queuewait
+
+import (
+	"context"
+	"time"
+
+	utilsclock "k8s.io/utils/clock"
+)
+
+// NewQueueWaitContextFactory returns a new instance of queueWaitContextFactory
+func NewQueueWaitContextFactory(clock utilsclock.PassiveClock) *queueWaitContextFactory {
+	return &queueWaitContextFactory{
+		clock:    clock,
+		newCtxFn: context.WithTimeout,
+	}
+}
+
+type queueWaitContextFactory struct {
+	clock utilsclock.PassiveClock
+
+	// comes handy to write unit test
+	newCtxFn func(parent context.Context, timeout time.Duration) (context.Context, context.CancelFunc)
+}
+
+// GetQueueWaitContext returns a new context with a deadline of how
+// long the request is allowed to wait before it is removed from its
+// queue and rejcted.
+// The context.CancelFunc returned must never be nil and the caller is
+// responsible for calling the CancelFunc function for cleanup.
+//   - ctx: the context associated with the request (it may or may
+//     not have a deadline).
+//   - defaultQueueWaitTime: the default wait duration that is used if
+//     the request context does not have any deadline.
+func (qw *queueWaitContextFactory) GetQueueWaitContext(parent context.Context, defaultQueueWaitTime time.Duration) (context.Context, context.CancelFunc) {
+	if parent.Err() != nil {
+		return parent, func() {}
+	}
+
+	if deadline, ok := parent.Deadline(); ok {
+		// we will allow the request to wait in the queue for one fourth of the time
+		// of its remaining deadline. (This is a new behavior).
+		// if the request context does not have any deadline then we default to
+		// 'defaultQueueWaitTime' which is 15s today. (This is in keeping with
+		// the current behavior)
+		queueWaitTime := deadline.Sub(qw.clock.Now()) / 4
+		if queueWaitTime <= 0 {
+			return parent, func() {}
+		}
+		return qw.newCtxFn(parent, queueWaitTime)
+	}
+
+	return qw.newCtxFn(parent, defaultQueueWaitTime)
```

Comment:
If defaultQueueWaitTime would be smaller than 1/4th of deadline, should we choose that then? Or do we ignore it completely?

Reply:
@wojtek-t so the default queue wait time will apply to requests without a context deadline, today this would be the watch requests, all other non-long running requests would have a deadline of `60s` or the specified `timeout` value in the request URI.

if a request specifies `timeout=10s` in the request URI, this PR will allow this request to sit in the queue for at most `2.5s` before it is rejected.

if a user specifies a value `>60s` in the `timeout` parameter, we ignore it and clamp it to `60s`, so 1/4th of the deadline should not exceed `15s`

I am inclined to keep the default queue wait time independent from the queue wait time derived from the request deadline, but happy to hear if you have any suggestions.

Also, are we okay with the 1/4 of the deadline as the queue wait time, or do we want to allow requests to wait in the queue for more time? 
please note this is a new behavior, previously a request with a `20s` timeout would wait in the queue for `15s`, now it will be rejected after `5s`